engine/src/test/java/org/camunda/bpm/engine/test/api/runtime/VariableInstanceTest.java

```diff
@@ -0,0 +1,178 @@
+    public void shouldUpdateVariableStateOnVariableTypeChangeStringToObject() {
+        // given
+        var processInstance = runtimeService.startProcessInstanceByKey("oneTaskProcess",
+                "variableA", Map.of("variableA", "This is a string value")
+        );
+
+        VariableInstanceEntity variable = (VariableInstanceEntity) runtimeService.createVariableInstanceQuery()
+                .processInstanceIdIn(processInstance.getId())
+                .singleResult();
+
+        // Double variable state is populated
+        assertThat(variable.getDoubleValue()).isNull();
+        assertThat(variable.getLongValue()).isNull();
+        assertThat(variable.getTextValue()).isEqualTo("This is a string value");
+        assertThat(variable.getTextValue2()).isNull();
+        assertThat(variable.getValue()).isEqualTo("This is a string value");
+
+        // The other type fields will be null
+        assertThat(variable.getByteArrayValue()).isNull();
+        assertThat(variable.getByteArrayValueId()).isNull();
+
+        // when the type is updated from double to object
+        setVariableWithObjectValue(processInstance.getId(), "variableA", "43.0"); // object value
+
+        variable = (VariableInstanceEntity) runtimeService.createVariableInstanceQuery()
+                .processInstanceIdIn(processInstance.getId())
+                .singleResult();
+
+        // then the changed object type should be reflected in the variable entity row appropriately
+        assertThat(variable.getByteArrayValue()).isNotNull();  // byte array fields should note exist for an integer field
+        assertThat(variable.getByteArrayValueId()).isNotNull();
+
+        assertThat(variable.getDoubleValue()).isNull();
+        assertThat(variable.getLongValue()).isNull();
+        assertThat(variable.getTextValue()).isNull();
+        assertThat(variable.getTextValue2()).isEqualTo("java.lang.String");
+    }
+
+    private ProcessInstance startProcessInstanceWithVariable(String processDefinitionKey, String variableName, Object variableValue) {
+        VariableMap variables = Variables.createVariables()
+                .putValue(variableName, variableValue);
+
+        return runtimeService.startProcessInstanceByKey(processDefinitionKey, variables);
+    }
+
+    private ProcessInstance startProcessInstanceWithObjectVariable(String processDefinitionKey, String variableName, Object variableValue) {
+        ObjectValue value = Variables.objectValue(variableValue)
+                .serializationDataFormat(Variables.SerializationDataFormats.JAVA)
+                .create();
+
+        VariableMap variables = Variables.createVariables()
+                .putValue(variableName, value);
+
+        String businessKey = processDefinitionKey; // for simplicity’s sake, same businessKey with processDefinitionKey
+        return runtimeService.startProcessInstanceByKey(processDefinitionKey, businessKey, variables);
```

Comment:
🙃 no need for business key:
```suggestion
        return runtimeService.startProcessInstanceByKey(processDefinitionKey, variables);
```